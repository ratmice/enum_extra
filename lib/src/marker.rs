pub trait NonZeroRepr {
    // TODO add an associated type
    // I guess that would technically not be a marker any longer
    // type NonZeroRepr;
}

#[cfg(test)]
#[allow(unused)]
mod test {
    use super::*;
    use enum_extra_derive::NonZeroRepr;
    use strum::EnumMetadata;
    use strum_macros::EnumMetadata;

    #[derive(NonZeroRepr, EnumMetadata)]
    enum XYZ {
        X = 24,
        Y = 25,
        Z = 26,
    }

    #[test]
    fn test_literal_discriminant() {
        assert!(generic_derived(XYZ::X))
    }

    fn generic_derived<X: NonZeroRepr>(_foo: X) -> bool {
        true
    }

    #[derive(NonZeroRepr)]
    enum XYZZ {
        X = 1 << 0,
        Y = 1 << 1,
        Z = 1 << 2,
    }

    #[test]
    fn test_expr_discriminant() {
        assert!(generic_derived(XYZZ::X))
    }
}
