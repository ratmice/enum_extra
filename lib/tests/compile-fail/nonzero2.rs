#[allow(unused_imports)]
use enum_extra::marker::NonZeroRepr;
use enum_extra_derive::NonZeroRepr;
use strum::EnumMetadata;
use strum_macros::EnumMetadata;

#[derive(EnumMetadata, NonZeroRepr)]
enum A {
    A = 0 << 1,
}

fn main() {
}


