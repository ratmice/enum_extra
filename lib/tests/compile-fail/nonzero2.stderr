error[E0080]: evaluation of constant value failed
 --> tests/compile-fail/nonzero2.rs:7:24
  |
7 | #[derive(EnumMetadata, NonZeroRepr)]
  |                        ^^^^^^^^^^^ the evaluated program panicked at 'Expected non-zero discriminant', $DIR/tests/compile-fail/nonzero2.rs:7:24
  |
  = note: this error originates in the macro `$crate::panic::panic_2015` (in Nightly builds, run with -Z macro-backtrace for more info)
